flatpak-import.py

```diff
@@ -0,0 +1,212 @@
+#!/usr/bin/env python2
+
+from argparse import ArgumentParser
+from ConfigParser import ConfigParser
+import gi
+gi.require_version('OSTree', '1.0')
+from gi.repository import GLib, Gio, OSTree
+import logging
+import os
+from sys import exit
+
+
+OSTREE_COMMIT_GVARIANT_STRING = "(a{sv}aya(say)sstayay)"
+
+OSTREE_STATIC_DELTA_META_ENTRY_FORMAT = "(uayttay)"
+
+OSTREE_STATIC_DELTA_FALLBACK_FORMAT = "(yaytt)"
+
+OSTREE_STATIC_DELTA_SUPERBLOCK_FORMAT = \
+    "(a{sv}tayay" + OSTREE_COMMIT_GVARIANT_STRING + \
+    "aya" + OSTREE_STATIC_DELTA_META_ENTRY_FORMAT + \
+    "a" + OSTREE_STATIC_DELTA_FALLBACK_FORMAT + \
+    ")"
+
+
+if __name__ == "__main__":
+
+    aparser = ArgumentParser(
+        description='Import flatpak into a local repository ',
+    )
+    aparser.add_argument('repo',
+                         help='repository name to use')
+    aparser.add_argument('flatpak', help='file to import')
+    aparser.add_argument('-g', '--gpg-homedir',
+                         help='GPG homedir to use when looking for keyrings')
+    aparser.add_argument('-k', '--keyring', help='additional trusted keyring file')
+    aparser.add_argument('-s', '--sign-key', help='GPG key ID to sign the commit with')
+    aparser.add_argument('-v', '--verbose', action='store_true',
+                         help='verbose log output')
+    aparser.add_argument('-d', '--debug', action='store_true',
+                         help='debug log output')
+
+    config = ConfigParser()
+    config.read('flatpak-import.conf')
+    aparser.set_defaults(**dict(config.items('defaults')))
+
+    args = aparser.parse_args()
+
+    if args.debug:
+        logging.basicConfig(level=logging.DEBUG)
+    elif args.verbose:
+        logging.basicConfig(level=logging.INFO)
+    else:
+        logging.basicConfig(level=logging.WARNING)
+
+
+    ### mmap the flatpak file and create a GLib.Variant from it
+
+    mapped_file = GLib.MappedFile.new(args.flatpak, False)
+    mapped_bytes = mapped_file.get_bytes()
+    ostree_static_delta_superblock_format = GLib.VariantType(
+            OSTREE_STATIC_DELTA_SUPERBLOCK_FORMAT)
+    delta = GLib.Variant.new_from_bytes(
+            ostree_static_delta_superblock_format,
+            mapped_bytes,
+            False)
+
+
+    ### parse flatpak metadata
+
+    # use get_child_value instead of array index to avoid
+    # slowdown (constructing the whole array?)
```

Comment:
Is it really slow? I was just thinking that the issue was that indexing the array gives you back a native python object rather than a GLib.Variant. You'd have to convert that before passing it to `validate_structureof_csum_v`. But whatever, this is fine.